tests/compatibility/forward_test.go

```diff
@@ -0,0 +1,323 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compatibility_test
+
+import (
+	"context"
+	"math/rand"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	pd "github.com/tikv/pd/client"
+	bs "github.com/tikv/pd/pkg/basicserver"
+	"github.com/tikv/pd/pkg/core"
+	"github.com/tikv/pd/pkg/mcs/discovery"
+	tso "github.com/tikv/pd/pkg/mcs/tso/server"
+	"github.com/tikv/pd/pkg/mcs/utils"
+	"github.com/tikv/pd/pkg/utils/tempurl"
+	"github.com/tikv/pd/pkg/utils/testutil"
+	"github.com/tikv/pd/tests"
+)
+
+// This part is to test the function of forwarding tso requests from the old client and the new server.
+// After splitting tso in mcs, tso requests sent to api server will be forwarded directly to tso server
+// in order to be compatible with the behavior of the old client. For the new client, it will connect
```

Comment:
IMPO, using failpoint to forcely use PD service mode instead of API mode might be enough.

Reply:
The current approach looks ok to me as well, because it can cover more cases and test the changes for tso mcs introducing any breaking change.

Reply:
> IMPO, using failpoint to forcely use PD service mode instead of API mode might be enough.

Maybe, the way with the old client is more similar to serverless env.

Reply:
If use failpoint, where do I need to add it?